src/SharpLearning.Backend.Cntk.Test/MnistCnnExampleTest.cs

```diff
@@ -92,12 +96,154 @@ public void MnistCnnTest()
+
+            // Training config.
+            uint minibatchSize = 64;
+            var minibatchIterations = 200;
+
+            // Instantiate progress writers.
+            var trainingProgressOutputFreq = 100;
+
+            // Instantiate the trainer object to drive the model training.
+            var lrSchedule = new TrainingParameterScheduleDouble(0.01, 1);
+            var learner = new List<Learner>() { Learner.SGDLearner(z.Parameters(), lrSchedule) };
+            var trainer = Trainer.CreateTrainer(z, loss, errorMetric, learner);
+
+            // Load train data.
+            var trainPath = Path.Combine(dataDirectoryPath, "Train-28x28_cntk_text.txt");
+            if (!File.Exists(trainPath))
+            { throw new FileNotFoundException("Data not present. Use MNIST CNTK python example to download data"); }
+
+            var readerTrain = CreateReader(trainPath,
+                epochSize: 60000,
+                inputDimensions: inputDimensions.Aggregate((d1, d2) => d1 * d2),
+                numberOfClasses: numOutputClasses,
+                randomize: false);
+
+            var featureStreamInfo = readerTrain.StreamInfo(FeatureId);
+            var labelStreamInfo = readerTrain.StreamInfo(LabelsId);
+
+            // Train model.
+            for (int i = 0; i < minibatchIterations; i++)
+            {
+                var mb = readerTrain.GetNextMinibatch(minibatchSize, device);
+                var arguments = new Dictionary<Variable, MinibatchData>
+                {
+                    { inputVar, mb[featureStreamInfo] },
+                    { labelVar, mb[labelStreamInfo] }
+                };
+                trainer.TrainMinibatch(arguments, device);
+
+                if (((i + 1) % trainingProgressOutputFreq) == 0 && trainer.PreviousMinibatchSampleCount() != 0)
+                {
+                    var trainLossValue = trainer.PreviousMinibatchLossAverage();
+                    var evaluationValue = trainer.PreviousMinibatchEvaluationAverage();
+                    Trace.WriteLine($"Minibatch: {i + 1} CrossEntropyLoss = {trainLossValue:F16}, EvaluationCriterion = {evaluationValue:F16}");
+                }
+            }
+
+            // Test model.
+            var mnist = Mnist.Load(DownloadPath);
+
+            var csharpError = MnistSimpleExampleTest.TestModelMnistLoader(z, device, mnist);
+            Trace.WriteLine($"Test Error: {csharpError}");
+
+            // Save model.
+            var modelPath = "lr_mnist_csharp_loader.dnn";
+            z.Save(modelPath);
+
+            // Test loaded model.
+            var loadedModel = Function.Load(modelPath, device);
+            var loadedModelError = MnistSimpleExampleTest.TestModelMnistLoader(z, device, mnist);
+
+            Trace.WriteLine("Loaded Model Error: " + loadedModelError);
+            Assert.AreEqual(csharpError, loadedModelError, 0.00001);
+
+            // Test against python example.
+            var pythonError = 0.884;
```

Comment:
@nietras C# and Python cntk cnn examples now return the same error. Examples still need some cleanup but at least they are deterministic and return the same test error. This was only possible using CPU and the uniform initializer.